astropy/utils/decorators.py

```diff
@@ -273,6 +275,245 @@ def delete(self):
+
+    Raises
+    ------
+    TypeError
+        If the new argument name cannot be found in the function
+        signature and arg_in_kwargs was False or if it is used to
+        deprecate the name of the ``*args``-, ``**kwargs``-like arguments.
+        At runtime such an Error is raised if both the new_name
+        and old_name were specified when calling the function and
+        "relax=False".
+
+    Notes
+    -----
+    The decorator should be applied to a function where the **name**
+    of an argument was changed but it applies the same logic.
+
+    Examples
+    --------
+    The deprecation warnings are not shown in the following examples.
+
+    To deprecate a positional or keyword argument::
+
+        >>> from astropy.utils.decorators import deprecated_renamed_argument
+        >>> @deprecated_renamed_argument('sig', 'sigma', '1.0')
+        ... def test(sigma):
+        ...     return sigma
+
+        >>> test(2)
+        2
+        >>> test(sigma=2)
+        2
+        >>> test(sig=2)
+        2
+
+    To deprecate an argument catched inside the ``**kwargs`` the
+    ``arg_in_kwargs`` has to be set::
+
+        >>> @deprecated_renamed_argument('sig', 'sigma', '1.0',
+        ...                             arg_in_kwargs=True)
+        ... def test(**kwargs):
+        ...     return kwargs['sigma']
+
+        >>> test(sigma=2)
+        2
+        >>> test(sig=2)
+        2
+
+    By default providing the new and old keyword will lead to an Exception. If
+    a Warning is desired set the ``relax`` argument::
+
+        >>> @deprecated_renamed_argument('sig', 'sigma', '1.0', relax=True)
+        ... def test(sigma):
+        ...     return sigma
+
+        >>> test(sig=2)
+        2
+
+    It is also possible to replace multiple arguments. The ``old_name``,
+    ``new_name`` and ``since`` have to be `tuple` or `list` and contain the
+    same number of entries::
+
+        >>> @deprecated_renamed_argument(['a', 'b'], ['alpha', 'beta'],
+        ...                              ['1.0', 1.2])
+        ... def test(alpha, beta):
+        ...     return alpha, beta
+
+        >>> test(a=2, b=3)
+        (2, 3)
+
+    In this case ``arg_in_kwargs`` and ``relax`` can be a single value (which
+    is applied to all renamed arguments) or must also be a `tuple` or `list`
+    with values for each of the arguments.
+
+    .. warning::
+        This decorator needs to access the original signature of the decorated
+        function. Therefore this decorator must be the **first** decorator on
+        any function if it needs to work for Python before version 3.4.
+    """
+    cls_iter = (list, tuple)
+    if isinstance(old_name, cls_iter):
+        n = len(old_name)
+        # Assume that new_name and since are correct (tuple/list with the
+        # appropriate length) in the spirit of the "consenting adults". But the
+        # optional parameters may not be set, so if these are not iterables
+        # wrap them.
```

Comment:
Especially because it's likely to fail nevertheless:

- It's unlikely that single letters of the `old_name` are part of the signature.
- It's likely that the "add sphinx directive" check fails if parts of the "since" are used.

So it would just be a "more accurate error message" if I added these checks. OTOH what if `new_name` is a `str` but `old_name` and `since` are lists. That case would also fail later (because it doesn't enter the if branch here). It would require a serious amount of additional checks here to make it totally safe (and slower).